mobile/android/android-device.ts

```diff
@@ -185,9 +188,166 @@ export class AndroidDevice implements Mobile.IDevice {
+        this.$childProcess.exec(installCommand).wait();
+
+        var port = options["debug-port"];
+
+        var clearDebugEnvironmentCommand = this.composeCommand('shell rm "/sdcard/Android/data/%s/files/envDebug.out"', packageName);
+        this.$childProcess.exec(clearDebugEnvironmentCommand).wait();
+
+        this.startPackageOnDevice(packageName).wait();
+
+        var dbgPort = this.startAndGetPort(packageName).wait();
+
+        if (dbgPort > 0) {
+            this.tcpForward(dbgPort, dbgPort).wait();
+            this.startDebuggerClient(dbgPort).wait();
+            this.openDebuggerClient(this.defaultNodeInspectorUrl + "?port=" + dbgPort).wait();
+        }
+    }
+
+    public debug(packageFile: string, packageName: string): IFuture<void> {
+        return (() => {
+            if (options["get-port"]) {
+                this.printDebugPort(packageName);
+            } else if (options["start"]) {
+                this.attachDebugger(packageName);
+            } else if (options["stop"]) {
+                this.detachDebugger(packageName);
+            } else if (options["debug-brk"]) {
+                this.startAppWithDebugger(packageFile, packageName);
+            } else {
+                this.$logger.info("Should specify at least one option: debug-brk, start, stop, get-port.");
+            }
+        }).future<void>()();
+    }
+
+    private checkIfRunning(packageName: string): boolean {
+        var isRunning = this.checkIfFileExists(packageName, "envDebug.out").wait();
+        return isRunning;
+    }
+
+    private readDebugPort(packageName: string): IFuture<number> {
+        return (() => {
+            var res = this.$childProcess.spawnFromEvent(this.adb, ["shell", "cat", "/sdcard/Android/data/" + packageName + "/files/envDebug.out"], "exit").wait();
+            var isRunning = res.stdout.indexOf('yes') > -1;
+            return 1;
+        }).future<number>()();
+    }
+
+    private sleep(ms: number): void {
+        var fiber = Fiber.current;
+        setTimeout(() => fiber.run(), ms);
+        Fiber.yield();
+    }
+
+    private checkIfFileExists(packageName: string, filename: string): IFuture<boolean> {
+        return (() => {
+            var args = ["shell", "test", "-f", "/sdcard/Android/data/" + packageName + "/files/" + filename, "&&", "echo 'yes'", "||", "echo 'no'"];
+            var res = this.$childProcess.spawnFromEvent(this.adb, args, "exit").wait();
+            var exists = res.stdout.indexOf('yes') > -1;
+            return exists;
+        }).future<boolean>()();
+    }
+
+    private startAndGetPort(packageName: string): IFuture<number> {
+        return (() => {
+            var port = -1;
+
+            var clearDebugEnvironmentCommand = this.composeCommand('shell rm "/sdcard/Android/data/%s/files/envDebug.in"', packageName);
+            this.$childProcess.exec(clearDebugEnvironmentCommand).wait();
+
+            var isRunning = false;
+            for (var i = 0; i < 10; i++) {
```

Comment:
The main intent was to emulate polling. I set the timeout for 10 second. These 10 seconds are distributed in a a loop with 1 second wait. Hence, 10x1000 = 10,000 ms = 10 seconds provide some responsiveness (it is not needed to wait all the 10 seconds if there is a response). Observable pattern is the right choice but adb tool does not provide it.